# BRIEF PREDICTION DE LA GRAVITE DES ACCIDENTS DE LA ROUTE

## Import des librairies

In [4]:
import numpy as np
import pandas as pd

import plotly.express as px
import matplotlib.pyplot as plt
import seaborn as sns

## Usagers

### Import du jeu de données 

In [7]:
df_usagers =pd.read_csv("usagers-2023.csv", delimiter = ';', encoding = 'utf_8')

In [8]:
df_usagers

Num_Acc    id_usager  id_vehicule num_veh  place  catu  grav  \
0       202300000001  203 851 184  155 680 557     A01      1     1     4   
1       202300000002  203 851 182  155 680 556     A01      1     1     1   
2       202300000002  203 851 183  155 680 556     A01     10     3     3   
3       202300000003  203 851 180  155 680 554     B01      1     1     3   
4       202300000003  203 851 181  155 680 555     A01      1     1     1   
...              ...          ...          ...     ...    ...   ...   ...   
125784  202300054821  203 720 720  155 583 346     A01      1     1     4   
125785  202300054821  203 720 721  155 583 347     B01      1     1     1   
125786  202300054822  203 720 717  155 583 344     A01      2     2     1   
125787  202300054822  203 720 718  155 583 344     A01      1     1     1   
125788  202300054822  203 720 719  155 583 345     B01      1     1     3   

        sexe  an_nais  trajet  secu1  secu2  secu3  locp actp  etatp  
0          1   1978.0       5      2      0     -1    -1   -1     -1  
1          2   1997.0       9      1      0     -1    -1   -1     -1  
2          1   1997.0       9      0     -1     -1     2    3      1  
3          1   1987.0       0      2      6      0     0    0     -1  
4          2   1984.0       0      1      0      0     0    0     -1  
...      ...      ...     ...    ...    ...    ...   ...  ...    ...  
125784     1   1971.0       1      2      0     -1     0    0     -1  
125785     2   1968.0       1      1      0     -1     0    0     -1  
125786     2   2003.0       2      1     -1     -1    -1   -1     -1  
125787     2   2002.0       1      1     -1     -1    -1   -1     -1  
125788     2   1995.0       1      2     -1     -1    -1   -1     -1  

[125789 rows x 16 columns]

In [9]:
df_usagers.head(20)

Num_Acc    id_usager  id_vehicule num_veh  place  catu  grav  sexe  \
0   202300000001  203 851 184  155 680 557     A01      1     1     4     1   
1   202300000002  203 851 182  155 680 556     A01      1     1     1     2   
2   202300000002  203 851 183  155 680 556     A01     10     3     3     1   
3   202300000003  203 851 180  155 680 554     B01      1     1     3     1   
4   202300000003  203 851 181  155 680 555     A01      1     1     1     2   
5   202300000004  203 851 175  155 680 551     B01      2     2     1     2   
6   202300000004  203 851 176  155 680 551     B01      1     1     1     2   
7   202300000004  203 851 177  155 680 552     C01      1     1     1     1   
8   202300000004  203 851 178  155 680 552     C01      2     2     4     1   
9   202300000004  203 851 179  155 680 553     A01      1     1     1     1   
10  202300000005  203 851 173  155 680 549     B01      1     1     4     2   
11  202300000005  203 851 174  155 680 550     A01      1     1     1     1   
12  202300000006  203 851 171  155 680 548     A01     10     3     4     1   
13  202300000006  203 851 172  155 680 548     A01      1     1     1     1   
14  202300000007  203 851 167  155 680 545     C01      1     1     1     1   
15  202300000007  203 851 168  155 680 546     A01      4     2     4     2   
16  202300000007  203 851 169  155 680 546     A01      1     1     1     1   
17  202300000007  203 851 170  155 680 547     B01      1     1     1     2   
18  202300000008  203 851 164  155 680 543     B01      4     2     4     2   
19  202300000008  203 851 165  155 680 543     B01      1     1     1     1   

    an_nais  trajet  secu1  secu2  secu3  locp actp  etatp  
0    1978.0       5      2      0     -1    -1   -1     -1  
1    1997.0       9      1      0     -1    -1   -1     -1  
2    1997.0       9      0     -1     -1     2    3      1  
3    1987.0       0      2      6      0     0    0     -1  
4    1984.0       0      1      0      0     0    0     -1  
5    2001.0       1      1      0     -1    -1   -1     -1  
6    1995.0       1      1      0     -1    -1   -1     -1  
7    1968.0       9      1      0     -1    -1   -1     -1  
8    1956.0       0      1      0     -1    -1   -1     -1  
9    1985.0       1      1      0     -1    -1   -1     -1  
10   2001.0       0      1      0     -1    -1   -1     -1  
11   2001.0       3      1      0     -1    -1   -1     -1  
12   2003.0       9      0     -1     -1     3    3      1  
13   1962.0       5      1      0     -1    -1   -1     -1  
14   1988.0       0      1      0     -1    -1   -1     -1  
15   1975.0       5      1      0     -1    -1   -1     -1  
16   1968.0       5      1      0     -1    -1   -1     -1  
17   1979.0       0      1      0     -1    -1   -1     -1  
18   2020.0       0      1      8     -1    -1   -1     -1  
19   1980.0       0      1      0     -1    -1   -1     -1

In [10]:
df_usagers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125789 entries, 0 to 125788
Data columns (total 16 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Num_Acc      125789 non-null  int64  
 1   id_usager    125789 non-null  object 
 2   id_vehicule  125789 non-null  object 
 3   num_veh      125789 non-null  object 
 4   place        125789 non-null  int64  
 5   catu         125789 non-null  int64  
 6   grav         125789 non-null  int64  
 7   sexe         125789 non-null  int64  
 8   an_nais      123191 non-null  float64
 9   trajet       125789 non-null  int64  
 10  secu1        125789 non-null  int64  
 11  secu2        125789 non-null  int64  
 12  secu3        125789 non-null  int64  
 13  locp         125789 non-null  int64  
 14  actp         125789 non-null  object 
 15  etatp        125789 non-null  int64  
dtypes: float64(1), int64(11), object(4)
memory usage: 15.4+ MB


In [11]:
df_usagers.describe()

Num_Acc          place           catu           grav  \
count  1.257890e+05  125789.000000  125789.000000  125789.000000   
mean   2.023000e+11       2.091646       1.332684       2.514544   
std    1.582901e+04       2.582452       0.611056       1.377953   
min    2.023000e+11      -1.000000       1.000000      -1.000000   
25%    2.023000e+11       1.000000       1.000000       1.000000   
50%    2.023000e+11       1.000000       1.000000       3.000000   
75%    2.023000e+11       2.000000       2.000000       4.000000   
max    2.023001e+11      10.000000       3.000000       4.000000   

                sexe        an_nais         trajet          secu1  \
count  125789.000000  123191.000000  125789.000000  125789.000000   
mean        1.274134    1984.252023       3.128692       1.932323   
std         0.561198      19.099389       2.754842       2.371647   
min        -1.000000    1913.000000      -1.000000      -1.000000   
25%         1.000000    1971.000000       0.000000       1.000000   
50%         1.000000    1988.000000       4.000000       1.000000   
75%         2.000000    2000.000000       5.000000       2.000000   
max         2.000000    2023.000000       9.000000       9.000000   

               secu2          secu3           locp          etatp  
count  125789.000000  125789.000000  125789.000000  125789.000000  
mean        0.902885      -0.885284      -0.236046      -0.816232  
std         2.980147       1.004342       1.279765       0.640256  
min        -1.000000      -1.000000      -1.000000      -1.000000  
25%        -1.000000      -1.000000      -1.000000      -1.000000  
50%         0.000000      -1.000000      -1.000000      -1.000000  
75%         0.000000      -1.000000       0.000000      -1.000000  
max         9.000000       9.000000       9.000000       3.000000

In [12]:
df_usagers.isnull().sum()

Num_Acc           0
id_usager         0
id_vehicule       0
num_veh           0
place             0
catu              0
grav              0
sexe              0
an_nais        2598
trajet            0
secu1             0
secu2             0
secu3             0
locp              0
actp              0
etatp             0
dtype: int64

In [13]:
df_usagers = df_usagers.dropna()

Jeu de données pour les usagers de 125789 lignes pour 16 colonnes.
Chaque ligne correspond à un usager différent donc nous pouvons supprimer la colonne "id_usager".
Un même accident ("Num_Acc") peut concerner plusieurs véhicules ("id_vehicule" et "num_veh") => on peut supprimer "num_veh"
Un même véhicule peut concerner plusieurs usagers.
Il peut y avoir jusqu'à 3 équipements de sécurité de renseignés par usager. 
On supprime les valeurs -1 qui ne sont pas attendues.

### Suppression des colonnes inutiles


In [16]:
df_usagers.drop(['id_usager'], axis = 1, inplace = True)

C:\Users\peggy\AppData\Local\Temp\ipykernel_24532\2549053182.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_usagers.drop(['id_usager'], axis = 1, inplace = True)


In [17]:
df_usagers.drop(['num_veh'], axis = 1, inplace = True)

C:\Users\peggy\AppData\Local\Temp\ipykernel_24532\3851949133.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_usagers.drop(['num_veh'], axis = 1, inplace = True)


In [18]:
df_usagers.head()

Num_Acc  id_vehicule  place  catu  grav  sexe  an_nais  trajet  secu1  \
0  202300000001  155 680 557      1     1     4     1   1978.0       5      2   
1  202300000002  155 680 556      1     1     1     2   1997.0       9      1   
2  202300000002  155 680 556     10     3     3     1   1997.0       9      0   
3  202300000003  155 680 554      1     1     3     1   1987.0       0      2   
4  202300000003  155 680 555      1     1     1     2   1984.0       0      1   

   secu2  secu3  locp actp  etatp  
0      0     -1    -1   -1     -1  
1      0     -1    -1   -1     -1  
2     -1     -1     2    3      1  
3      6      0     0    0     -1  
4      0      0     0    0     -1

### Observation des variables

In [20]:
# Variable 'place' => place dans la voiture

In [21]:
#Place occupée dans le véhicule
df_usagers['place'].value_counts(normalize = True)

place
 1     0.737521
 2     0.114091
 10    0.077230
 3     0.019636
 4     0.018427
 9     0.010618
 7     0.009887
 5     0.006202
 8     0.004732
 6     0.001640
-1     0.000016
Name: proportion, dtype: float64

In [22]:
# Réduction du nombre de classe => On conserve le conducteur, le passager avant, les piétons liés à l'accident et les 2 passagers arrières. 
# On regroupe les autres passagers (des véhicules 9 places par ex) en 1 seule catégorie "autre"
# Affectation directe sans chaine
df_usagers['place'] = df_usagers['place'].replace([5, 6, 7, 8, 9], 'autre')

C:\Users\peggy\AppData\Local\Temp\ipykernel_24532\4232736929.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_usagers['place'] = df_usagers['place'].replace([5, 6, 7, 8, 9], 'autre')


In [23]:
# Suppression des -1
df_usagers = df_usagers[df_usagers['place']!= -1]

In [24]:
df_usagers['place'].value_counts(normalize = True)

place
1        0.737533
2        0.114093
10       0.077231
autre    0.033079
3        0.019636
4        0.018427
Name: proportion, dtype: float64

In [25]:
# Catégorie d'usager
df_usagers['catu'].value_counts(normalize = True)

catu
1    0.737712
2    0.185057
3    0.077231
Name: proportion, dtype: float64

In [26]:
# Les proportions étant sensiblement les mêmes que pour la variable "place", on supprime la colonne "catu"
df_usagers.drop('catu', axis = 1, inplace = True)

In [27]:
# Gravité de blessures de l'usager
df_usagers['grav'].value_counts(normalize = True)

grav
1    0.413340
4    0.402641
3    0.156434
2    0.027584
Name: proportion, dtype: float64

In [28]:
# Target ? Reduction en 2 classes "grave" "pas grave" ??

In [29]:
df_usagers['sexe'].value_counts(normalize = True)

sexe
 1    0.680621
 2    0.319306
-1    0.000073
Name: proportion, dtype: float64

In [30]:
# Suppression des -1
df_usagers = df_usagers[df_usagers['sexe']!= -1]

In [31]:
# Année de naissance
df_usagers['an_nais'].unique()

array([1978., 1997., 1987., 1984., 2001., 1995., 1968., 1956., 1985.,
       2003., 1962., 1988., 1975., 1979., 2020., 1980., 1981., 1999.,
       1953., 2010., 1977., 1996., 1982., 1994., 1992., 2022., 1998.,
       2004., 1974., 1958., 2002., 2015., 2011., 1990., 2007., 2009.,
       2008., 1991., 1967., 2000., 2006., 1955., 1969., 1963., 1946.,
       1973., 1947., 1942., 1989., 1964., 1993., 1986., 1957., 1961.,
       1932., 1929., 1972., 2016., 1928., 1965., 1943., 1971., 1983.,
       1970., 1960., 2014., 1976., 1939., 2005., 1950., 1945., 1959.,
       1952., 1951., 1941., 1954., 2019., 1937., 1936., 1934., 1944.,
       1966., 2017., 1940., 1948., 2012., 2023., 2013., 1949., 1933.,
       2018., 1935., 1930., 2021., 1931., 1938., 1927., 1923., 1926.,
       1924., 1925., 1921., 1922., 1917., 1920., 1919., 1913.])

In [32]:
# On fait des classes pour les années de naissance

In [33]:
for i, row in df_usagers.iterrows() :
    if row['an_nais'] <= 1953 :
        df_usagers.at[i,'années'] = "1910 - 1953"
    elif row['an_nais'] <= 1973 :
        df_usagers.at[i,'années'] = "1954 - 1973"
    elif row['an_nais'] <= 1999 :
        df_usagers.at[i,'années'] = "1974 - 1999"
    elif row['an_nais'] >= 2000 :
        df_usagers.at[i,'années'] = "2000 - 2023"
    
    


In [34]:
df_usagers['années'].unique()

array(['1974 - 1999', '2000 - 2023', '1954 - 1973', '1910 - 1953'],
      dtype=object)

In [35]:
# Motif du déplacement
df_usagers['trajet'].value_counts(normalize = True)

trajet
 5    0.374452
 0    0.268566
 1    0.133731
 4    0.089260
 9    0.079323
 3    0.029193
 2    0.024241
-1    0.001234
Name: proportion, dtype: float64

In [36]:
# les motifs de déplacement non connus (classe 0, -1 et 9) représentent 36% 
# Suppression de la colonne
df_usagers.drop('trajet', axis = 1, inplace = True)

In [37]:
# Equipement de sécurité
df_usagers['secu1'].value_counts(normalize = True)

secu1
 1    0.591338
 2    0.178519
 8    0.122569
 0    0.096785
 3    0.005894
-1    0.002038
 6    0.001088
 9    0.000999
 4    0.000544
 5    0.000219
 7    0.000008
Name: proportion, dtype: float64

In [38]:
# Réduction du nombre de classe
df_usagers['secu1'] = df_usagers['secu1'].replace(8, -1)
df_usagers['secu1'] = df_usagers['secu1'].replace([3, 4, 5, 6, 7], 9)

In [39]:
# Suppression des -1
df_usagers['secu1'] = df_usagers['secu1'].replace([-1,9], 8)

In [40]:
df_usagers['secu1'].value_counts(normalize = True)

secu1
1    0.591338
2    0.178519
8    0.133358
0    0.096785
Name: proportion, dtype: float64

In [41]:
df_usagers['secu2'].value_counts(normalize = True)

secu2
-1    0.431125
 0    0.361341
 6    0.096444
 8    0.079372
 4    0.012380
 5    0.009368
 9    0.003791
 1    0.002509
 3    0.001486
 7    0.001145
 2    0.001039
Name: proportion, dtype: float64

In [42]:
# Suppression de la colonne car trop d'inconnues
df_usagers.drop('secu2', axis = 1, inplace = True)

In [43]:
df_usagers['secu3'].value_counts(normalize = True)

secu3
-1    0.981702
 9    0.008305
 0    0.006219
 1    0.001616
 8    0.000909
 6    0.000649
 2    0.000195
 4    0.000179
 5    0.000138
 3    0.000065
 7    0.000024
Name: proportion, dtype: float64

In [44]:
# Suppression de la colonne car trop d'inconnues
df_usagers.drop('secu3', axis = 1, inplace = True)

In [45]:
df_usagers['locp'].value_counts(normalize = True)

locp
-1    0.495024
 0    0.423056
 3    0.026522
 2    0.016042
 4    0.012413
 1    0.011796
 5    0.006819
 9    0.004278
 6    0.002728
 8    0.001234
 7    0.000089
Name: proportion, dtype: float64

In [46]:
# Suppression de la colonne car trop d'inconnues
df_usagers.drop('locp', axis = 1, inplace = True)

In [47]:
df_usagers['actp'].value_counts(normalize = True)

actp
 -1    0.494407
0      0.423697
3      0.060367
9      0.005626
1      0.005317
B      0.003028
5      0.002809
2      0.002557
A      0.000958
4      0.000731
6      0.000235
7      0.000195
8      0.000073
Name: proportion, dtype: float64

In [48]:
# Suppression de la colonne car trop d'inconnues
df_usagers.drop('actp', axis = 1, inplace = True)

In [49]:
df_usagers['etatp'].value_counts(normalize = True)

etatp
-1    0.918112
 1    0.061577
 2    0.016967
 3    0.003345
Name: proportion, dtype: float64

In [50]:
# Suppression de la colonne car trop d'inconnues
df_usagers.drop('etatp', axis = 1, inplace = True)

In [51]:
df_usagers.shape


(123180, 8)

In [52]:
#  Suppression de 2609 lignes et de 8 colonnes

## Exportation du dataframe nettoyé

In [54]:
## Exportation du dataframe nettoyé
df_usagers.to_csv('usagers_clean.csv', index=False)